### PR TITLE
datatype: get ready for large count datatypes

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Turning every internal datatype creation routines to use `MPI_Aint` and add implementations for large count datatype creation APIs.

There is a yaksa PR https://github.com/pmodels/yaksa/pull/171 to update yaksa API to match the `MPI_Aint` internal types (using `intptr_t`). Currently, in this PR we have assertions and transcribe the counts-array before calling yaksa. Once yaksa is updated, those assertions and transcribe will be removed.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/sock

**Comment by hzhou** [comment]
test:mpich/ch4/most

File: src/mpi/datatype/typerep/src/typerep_dataloop_commit.c [hunk 1]
```diff
@@ -137,20 +136,22 @@ void MPIR_Typerep_commit(MPI_Datatype type)
...
-        case MPI_COMBINER_HINDEXED_INTEGER:
         case MPI_COMBINER_HINDEXED:
-        case MPI_COMBINER_STRUCT_INTEGER:
         case MPI_COMBINER_STRUCT:
             if (ints[0] == 0) {
                 MPIR_Dataloop_create_contiguous(0, MPI_INT, (void **) dlp_p);
                 goto clean_exit;
             }
             break;
+        case MPI_COMBINER_HVECTOR_INTEGER:
```
Comment (raffenet) [inline src/mpi/datatype/typerep/src/typerep_dataloop_commit.c:144]: Just thinking about the ABI implications of this change. Does an application always have to query for the combiner type? Could it hardcoded somehow?

File: src/include/mpir_datatype.h [hunk 2]
```diff
@@ -492,6 +492,43 @@ static inline int MPIR_Datatype_set_contents(MPIR_Datatype * new_dtp,
     return MPI_SUCCESS;
 }
 
+MPL_STATIC_INLINE_PREFIX void MPIR_Datatype_access_contents(MPIR_Datatype_contents * cp,
```
Comment (raffenet) [inline src/include/mpir_datatype.h:470]: Suggest `MPIR_Datatype_get_contents`.

Comment (raffenet) [reply src/include/mpir_datatype.h:470]: Nevermind, I see later in the history what is going on. Naming is still weird to me, but not your change.

Comment (hzhou) [reply src/include/mpir_datatype.h:470]: There is `MPI_Type_get_contents`, which return *copies* of the internal counts array. `MPIR_Datatype_access_contents` just returns the pointer to the internal data, I think the name distinction is important.

File: src/mpi/datatype/typerep/src/typerep_dataloop_commit.c (hunk 1 above)
Comment (hzhou) [reply src/mpi/datatype/typerep/src/typerep_dataloop_commit.c:144]: User do query the combiner with `MPI_Type_get_envelope`.

**Review by raffenet — COMMENTED** [review]
These changes look like they will impose overhead on non-large count operations by forcing promotion of arrays of `int` to `MPI_Aint`. My takeaway from some of our Slack discussion about this was that we weren't going to impose this kind of overhead on things that didn't actually need large count. Was I wrong? I guess this only impacts a subset of functionality, and none of it is for predefined types, so the actual application impact could still be quite small.

**Comment by hzhou** [comment]
> These changes look like they will impose overhead on non-large count operations by forcing promotion of arrays of `int` to `MPI_Aint`. My takeaway from some of our Slack discussion about this was that we weren't going to impose this kind of overhead on things that didn't actually need large count. Was I wrong? I guess this only impacts a subset of functionality, and none of it is for predefined types, so the actual application impact could still be quite small.

As I was implementing it, to use intermediate datatype is not trivial and it will introduce much extra complexity -- essentially when you review the code, you see extra code logic doing work-around, and in the case of non-even large blklens array, everything has to become a `struct`. Making internal large is much simpler and I believe more robust in the long run.

As for the performance, first of all, the common types -- contig and vector types -- are not affected as they are just scalar parameters. So we are talking about the impact on `indexed_block`, `indexed`, and `struct`. Then I believe the common case are with small counts array, in which the overhead of transcribing the counts are negligible compared to the datatype creation itself. 

So that only leaves *very rare* cases (admittedly not so rare in our test suite) where the user creates a datatype that has very large non-contig irregular segments. And if the transcription of the counts-array is an issue, I believe the user will have a much bigger performance issue when the user actually communicating with the datatype, and again, I believe the overhead of transcription comparatively is negligible.

So overall, I believe just make internal type large is a better solution.

**Comment by hzhou** [comment]
By the way, internally we will copy the counts-array to `contents` and `yaksa`, so we are doing the extra copy 1 or 2 times already. Having 1 extra copy for the rare cases should not have much visible impact comparatively.

**Comment by raffenet** [comment]
> As for the performance, first of all, the common types -- contig and vector types -- are not affected as they are just scalar parameters. So we are talking about the impact on `indexed_block`, `indexed`, and `struct`. Then I believe the common case are with small counts array, in which the overhead of transcribing the counts are negligible compared to the datatype creation itself.

Yes, this is what I was gathering when reading the yaksa PR. We are on the same page regarding overheads then. I'll continue with the review.

File: src/mpi/datatype/typerep/src/typerep_dataloop_commit.c (hunk 1 above)
Comment (raffenet) [reply src/mpi/datatype/typerep/src/typerep_dataloop_commit.c:144]: Yeah, I suppose this is truly runtime information. Properly written programs should not be affected.

**Review by raffenet — APPROVED** [review]
LGTM

**Pull request merged**